### PR TITLE
Custom WASM downloading & execution

### LINKED ISSUE #37: ✨ Implement the Data Request input
## ✨ Feature

Implement the Data Request input parameter and pass it to the VM

## Motivation

This allows the user to customise the Data Request and behave differently on the input

## Implementation

* Pass the input parameters as args inside of Wasmer

### PR DESCRIPTION
**Pull request by FranklinWaller**
Implementation of custom WASM downloading & execution.

## Motivation

This allows users to create Data Requests with custom WASM binaries.

## Explanation of Changes

* Implement gRPC transparent encoder for sending and receiving arbitrary data
* Implement protobuf
* Before executing a DR the node checks if the `dr_binary_id` exists. If not it will try to download the binary from the SEDA chain and then cache it locally. 
* If the binary exists it will execute the binary and wait for its result. Once the result is available the node will post the results on chain

## Testing

This assumes there is already a contract deployed and configured for the node. The node needs to be registered with the oracle contract.

1. Start the node
2. If the chain got reset you need to build and upload the simple-dr. See section to build and upload simple-dr
3. Go to https://codesandbox.io/s/web3-react-v6-forked-5wc3yq?file=/src/index.js to generate a new data request (make sure to change something in the DR to not generate duplicates, best is to add a byte to `memo`, which has no effect on the execution)
4. Look in the console and find the `post_data_request` JSON.  (`{"post_data_request":{"posted_dr":{"dr_id": ... }}}`)
5. Copy the JSON and go to Ping Pub Explorer
6. Go to Cosmwasm and find the oracle contract that you are currently using
7. Click execute and paste the JSON as args
8. Sign and execute

The node will now pickup the DR and execute it. It will take care of the downloading and executing

### Simple-dr building & uploading
NOTICE: If the chain didn't reset i've already upload the latest simple dr binary: `006fd965caffcb4ec8fac83a192ec8ddbe3da20083621d7f8866bed4086fdee4` you don't have to do this then.

1. Go to the folder `wasm/test/simple-dr`
2. Execute the command `cargo build --release --target wasm32-wasi`
10. Take the `simple-dr.wasm` from target/wasm32-wasi/release and move it to a location where you have access to the latest `seda-chaind` (Can be inside the `seda-chain` project)
11. Execute the command `./seda-chaind tx wasm-storage store-data-request-wasm ./simple-dr.wasm --wasm-type data-request --from <YOUR_ADDRESS> --node tcp://18.169.59.167:26657 --gas 9647711` Where ofcourse the `./simple-dr.wasm` is accessible aswell as `./seda-chaind`
12. On pingpub look for your transaction and in the logs find Something like: `{"key":"hash","value":"\"006fd965caffcb4ec8fac83a192ec8ddbe3da20083621d7f8866bed4086fdee4\""}`

## Related PRs and Issues

closes #37 #12

### CODE REVIEW

File: wasm/consensus/src/tasks/run.rs [hunk 1]
```diff
@@ -8,8 +8,8 @@ use crate::{
         trigger_eligibility_tick,
         trigger_execute_tick,
         trigger_fetch_tick,
-        trigger_relayer_tick,
-        trigger_relayer_to_sub_tick,
+        // trigger_relayer_tick,
+        // trigger_relayer_to_sub_tick,
```
Comment (Thomasvdam) [inline wasm/consensus/src/tasks/run.rs:12]: Why are we no longer using these functions? Is it a deliberate choice or was it something for testing/debugging?

File: wasm/test/simple-dr/src/main.rs [hunk 2]
```diff
@@ -1,39 +1,5 @@
...
-            println!("We've done it {response:?} {inner_body:?}");
-        }
-        PromiseStatus::Rejected(err) => {
-            let response = String::from_utf8_lossy(&err);
-            eprintln!("Oh noes {response:?}");
-        }
-    }
-
-    execution_result("ok".to_bytes());
+    execution_result("Hello World".to_bytes());
```
Comment (Thomasvdam) [inline wasm/test/simple-dr/src/main.rs:4]: Is this a legit test...?

File: wasm/consensus/src/tasks/execute/mod.rs [hunk 3]
```diff
@@ -2,7 +2,7 @@ use clap::Subcommand;
 
 use crate::errors::Result;
 
-mod execute_dr;
+pub mod execute_dr;
```
Comment (Thomasvdam) [inline wasm/consensus/src/tasks/execute/mod.rs:5]: What makes it so this needs to be public now?

**Review by Thomasvdam — COMMENTED** [review]
I'm unable to build the node locally so I'll leave that exercise to someone else, but most of the implementation looks sound to me. Just a few questions about some removed code (and the related dead code statements) .

File: Cargo.toml [hunk 4]
```diff
@@ -79,10 +80,11 @@ thiserror = "1.0"
...
 toml = "0.7"
+tonic = "0.9"
 tracing = { version = "0.1", features = ["log-always"] }
 tracing-appender = "0.2"
 tracing-subscriber = { version = "0.3", default-features = false }
 url = { version = "2.3", default-features = false }
-wasmer = { git = "https://github.com/wasmerio/wasmer.git", rev = "82974aca26b7d0a0a6f19f62a39a1d2ad07bdfad", default-features = true }
-wasmer-cache = { git = "https://github.com/wasmerio/wasmer.git", rev = "82974aca26b7d0a0a6f19f62a39a1d2ad07bdfad", default-features = true }
-wasmer-wasix = { git = "https://github.com/wasmerio/wasmer.git", rev = "82974aca26b7d0a0a6f19f62a39a1d2ad07bdfad", default-features = true }
+# wasmer 3.3.0
```
Comment (mariocao) [inline Cargo.toml:88]: Good catch!! 👍

File: Cargo.toml [hunk 5]
```diff
@@ -79,10 +80,11 @@ thiserror = "1.0"
...
+wasmer = { git = "https://github.com/wasmerio/wasmer.git", rev = "3c7c4a803e59d380ea969baa3f032598e41a1bcb", default-features = true }
+wasmer-wasix = { git = "https://github.com/wasmerio/wasmer.git", rev = "3c7c4a803e59d380ea969baa3f032598e41a1bcb", default-features = true }
```
Comment (mariocao) [inline Cargo.toml:90]: I suppose you already checked, but is there not any version that works for us? :/ 

I tried to check the commit in the repo, but it seems it does not belong to any branch.

File: .github/workflows/push.yml [hunk 6]
```diff
@@ -21,6 +21,12 @@ jobs:
           toolchain: nightly
           components: rustfmt
 
+      - name: Install Buf
+        uses: bufbuild/buf-setup-action@v1.26.1
+
+      - name: Build Protobuf
+        run: make build-proto
```
Comment (gluax) [inline .github/workflows/push.yml:28]: This is typically done using prost in a build.rs file.

**Review by gluax — APPROVED** [review]

File: .github/workflows/push.yml (hunk 6 above)
Comment (FranklinWaller) [reply .github/workflows/push.yml:28]: We can look into that later but the main reason was that it needs dependencies

File: Cargo.toml (hunk 5 above)
Comment (FranklinWaller) [reply Cargo.toml:90]: This commit is actually a tag taken from the releases. But no the version that was uploaded seems corrupted, and 4.0 is not compatible with our current structure sadly.

File: wasm/consensus/src/tasks/run.rs (hunk 1 above)
Comment (FranklinWaller) [reply wasm/consensus/src/tasks/run.rs:12]: It's temp because the relayer is currently broken (not compatible with the new EVM contract)

**Review by mariocao — APPROVED** [review]

**Pull request merged**